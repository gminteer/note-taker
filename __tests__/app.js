/* global shouldBeFound:writable, shouldValidate:writable, writeShouldSucceed:writable */
// using globals since I don't think there's an easy way to grab a reference to the mock Notes object the app has.
const request = require('supertest');

jest.mock('../lib/notes');
// eslint-disable-next-line no-unused-vars
const Notes = require('../lib/notes');
const app = require('../app');

beforeEach(() => {
  shouldBeFound = true;
  shouldValidate = true;
  writeShouldSucceed = true;
});

// API routes

describe('GET /api/notes', () => {
  test('should respond 200 with body containing notes array', async () => {
    const response = await request(app).get('/api/notes').expect(200);
    expect(response.body).toEqual([
      {title: 'testNote1', text: 'testText1', id: 0},
      {title: 'testNote2', text: 'testText2', id: 1},
      {title: 'testNote3', text: 'testText3', id: 2},
    ]);
  });
});

describe('GET /api/notes/:id', () => {
  test('should respond 200 with body containing a note if found on success', async () => {
    const response = await request(app).get('/api/notes/test').expect(200);
    expect(response.body).toEqual({title: 'testNote1', text: 'testText1', id: 0});
  });
  test('should respond 404 if not found', async () => {
    shouldBeFound = false;
    await request(app).get('/api/notes/test').expect(404);
  });
  test('should respond 500 if anything else goes wrong', async () => {
    writeShouldSucceed = false;
    await request(app).get('/api/notes/test').expect(500);
  });
});

describe('POST /api/notes', () => {
  test('should respond 201 with location header and body containing created note on success', async () => {
    const response = await request(app).post('/api/notes').send({test: 'test'}).expect(201);
    expect(response.header.location).toEqual('./0');
    expect(response.body).toEqual({test: 'test', id: 0});
  });
  test('should respond 400 if body fails validation', async () => {
    shouldValidate = false;
    await request(app).post('/api/notes').send({test: 'test'}).expect(400);
  });
  test('should respond 500 if write fails', async () => {
    writeShouldSucceed = false;
    await request(app).post('/api/notes').send({test: 'test'}).expect(500);
  });
});

describe('PUT /api/notes/:id', () => {
  test('should respond 200 with body containing updated note on success', async () => {
    const response = await request(app).put('/api/notes/test').send({test: 'test'}).expect(200);
    expect(response.body).toEqual({test: 'test'});
  });
  test('should respond 400 if body fails validation', async () => {
    shouldValidate = false;
    await request(app).put('/api/notes/test').send({test: 'test'}).expect(400);
  });
  test('should respond 404 if not found', async () => {
    shouldBeFound = false;
    await request(app).put('/api/notes/test').send({test: 'test'}).expect(404);
  });
  test('should respond 500 if write fails', async () => {
    writeShouldSucceed = false;
    await request(app).put('/api/notes/test').send({test: 'test'}).expect(500);
  });
});

describe('DELETE /api/notes/:id', () => {
  test('should respond 204 on success', async () => {
    await request(app).delete('/api/notes/test').expect(204);
  });
  test('should respond 404 if not found', async () => {
    shouldBeFound = false;
    await request(app).delete('/api/notes/test').expect(404);
  });
  test('should respond 500 if write fails', async () => {
    writeShouldSucceed = false;
    await request(app).delete('/api/notes/test').expect(500);
  });
});

// HTML routes
describe('GET /', () => {
  test('should respond 200', async () => {
    await request(app).get('/').expect(200);
  });
});
describe('GET /notes', () => {
  test('should respond 200', async () => {
    await request(app).get('/notes').expect(200);
  });
});
