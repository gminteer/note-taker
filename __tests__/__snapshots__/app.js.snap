// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DELETE /api/notes/:id should respond 500 with body containing error if data fails to initialize 1`] = `"init failed with \\"EMOCKERROR\\""`;

exports[`DELETE /api/notes/:id should respond 500 with body containing error if invalid id matches 1`] = `"\\"test\\" is invalid id, but found match"`;

exports[`DELETE /api/notes/:id should respond 500 with body containing error if write fails 1`] = `"write failed with \\"EMOCKERROR\\""`;

exports[`GET (anything else) should respond 200 with body containing static content 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

<head>
  <meta charset=\\"UTF-8\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
  <title>Note Taker</title>
  <link rel=\\"stylesheet\\" href=\\"https://cdnjs.cloudflare.com/ajax/libs/bootswatch/4.1.3/flatly/bootstrap.min.css\\" />
</head>

<body>
  <nav class=\\"navbar navbar-dark bg-dark\\">
    <a class=\\"navbar-brand\\" href=\\"/\\">Note Taker
    </a>
  </nav>
  <div class=\\"container\\">
    <div style=\\"margin-top: 80px;\\" class=\\"jumbotron text-center\\">
      <h1 class=\\"display-4\\">Note Taker <span role=\\"img\\" aria-label=\\"Memo\\">üìù</span></h1>
      <h4 class=\\"mt-4\\">Take notes with Express</h4>
      <a class=\\"btn btn-primary btn-lg mt-4\\" href=\\"/notes\\" role=\\"button\\">Get Started</a>
    </div>
  </div>
  <script src=\\"https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js\\"></script>
  <script src=\\"/assets/js/index.js\\"></script>
</body>

</html>
"
`;

exports[`GET / should respond 200 with body containing static content 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

<head>
  <meta charset=\\"UTF-8\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
  <title>Note Taker</title>
  <link rel=\\"stylesheet\\" href=\\"https://cdnjs.cloudflare.com/ajax/libs/bootswatch/4.1.3/flatly/bootstrap.min.css\\" />
</head>

<body>
  <nav class=\\"navbar navbar-dark bg-dark\\">
    <a class=\\"navbar-brand\\" href=\\"/\\">Note Taker
    </a>
  </nav>
  <div class=\\"container\\">
    <div style=\\"margin-top: 80px;\\" class=\\"jumbotron text-center\\">
      <h1 class=\\"display-4\\">Note Taker <span role=\\"img\\" aria-label=\\"Memo\\">üìù</span></h1>
      <h4 class=\\"mt-4\\">Take notes with Express</h4>
      <a class=\\"btn btn-primary btn-lg mt-4\\" href=\\"/notes\\" role=\\"button\\">Get Started</a>
    </div>
  </div>
  <script src=\\"https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js\\"></script>
  <script src=\\"/assets/js/index.js\\"></script>
</body>

</html>
"
`;

exports[`GET /api/notes should respond 200 with body containing notes array 1`] = `
Array [
  Object {
    "id": 0,
    "text": "testText1",
    "title": "testNote1",
  },
  Object {
    "id": 1,
    "text": "testText2",
    "title": "testNote2",
  },
  Object {
    "id": 2,
    "text": "testText3",
    "title": "testNote3",
  },
]
`;

exports[`GET /api/notes should respond 500 with body containing JSONified error if anything else goes wrong outside of production environment 1`] = `
Object {
  "code": "EMOCKERROR",
}
`;

exports[`GET /api/notes should respond 500 with body containing error if data fails to initialize 1`] = `"init failed with \\"EMOCKERROR\\""`;

exports[`GET /api/notes/:id should respond 200 with body containing a note if found 1`] = `
Object {
  "id": 0,
  "text": "testText1",
  "title": "testNote1",
}
`;

exports[`GET /api/notes/:id should respond 500 with body containing error if data fails to initialize 1`] = `"init failed with \\"EMOCKERROR\\""`;

exports[`GET /api/notes/:id should respond 500 with body containing error if invalid id matches 1`] = `"\\"test\\" is invalid id, but found match"`;

exports[`GET /notes should respond 200 with body containing static content 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

<head>
  <meta charset=\\"UTF-8\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
  <title>Note Taker</title>
  <link rel=\\"stylesheet\\" href=\\"https://cdnjs.cloudflare.com/ajax/libs/bootswatch/4.1.3/flatly/bootstrap.min.css\\" />
  <link rel=\\"stylesheet\\" href=\\"https://use.fontawesome.com/releases/v5.3.1/css/all.css\\" integrity=\\"sha384-mzrmE5qonljUremFsqc01SB46JvROS7bZs3IO2EmfFsd15uHvIt+Y8vEf7N7fWAU\\"
    crossorigin=\\"anonymous\\">
  <link rel=\\"stylesheet\\" href=\\"/assets/css/styles.css\\">
</head>

<body>
  <nav class=\\"navbar navbar-dark bg-dark\\">
    <a class=\\"navbar-brand\\" href=\\"/\\">Note Taker
    </a>
    <div class=\\"icons\\">
      <i class=\\"fas fa-save text-light save-note\\"></i>
      <i class=\\"fas fa-pen text-light new-note\\"></i>
    </div>
  </nav>
  <div class=\\"container-fluid\\">
    <div class=\\"row\\">
      <div class=\\"col-4 list-container\\">
        <div class=\\"card\\">
          <ul class=\\"list-group\\">
          </ul>
        </div>
      </div>
      <div class=\\"col-8\\">
        <input class=\\"note-title\\" placeholder=\\"Note Title\\" maxlength=\\"28\\" type=\\"text\\">
        <textarea class=\\"note-textarea\\" placeholder=\\"Note Text\\"></textarea>
      </div>
    </div>
  </div>
  <script src=\\"https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js\\"></script>
  <script src=\\"/assets/js/index.js\\"></script>
</body>

</html>
"
`;

exports[`POST /api/notes should respond 400 with body containing validation errors if body fails validation 1`] = `"suggested note does not validate"`;

exports[`POST /api/notes should respond 500 with body containing error if data fails to initialize 1`] = `"init failed with \\"EMOCKERROR\\""`;

exports[`POST /api/notes should respond 500 with body containing error if write fails 1`] = `"write failed with \\"EMOCKERROR\\""`;

exports[`PUT /api/notes/:id should respond 400 with body containing validation errors if body fails validation 1`] = `"suggested note does not validate"`;

exports[`PUT /api/notes/:id should respond 500 with body containing error if data fails to initialize 1`] = `"init failed with \\"EMOCKERROR\\""`;

exports[`PUT /api/notes/:id should respond 500 with body containing error if invalid id matches 1`] = `"\\"undefined\\" is invalid id, but found match"`;

exports[`PUT /api/notes/:id should respond 500 with body containing error if write fails 1`] = `"write failed with \\"EMOCKERROR\\""`;
